### PR TITLE
fix: table attribute cannot applied on td element

### LINKED ISSUE #6507: Bug: Table footer will not take tag attributes
### PHP Version

8.1

### CodeIgniter4 Version

4.1.5

### CodeIgniter4 Installation Method

Manual (zip or tar.gz)

### Which operating systems have you tested for this bug?

Linux

### Which server did you use?

apache

### Database

MySQL

### What happened?

**\CodeIgniter\View\Table**
Table footer will not incorporate extra attributes. It assumes the footer cell is `th` rather than `td` (as specified in the default template). This means you can not style the table footer by using element attributes (e.g. class, style).

### Steps to Reproduce

```php
// style wanted for cells containing numbers 
$style = 'background:#F99;text-align:right;width:5em;';
// table data
$data = [
	['item', ['data'=>'count', 'style'=>$style]],
	['shirt', ['data'=>2, 'style'=>$style]],
	['hat', ['data'=>1, 'style'=>$style]]
];
// add footing - summary of table data
$footing = ['total', ['data'=>3, 'style'=>$style]];

echo '<p>Use default template</p>';
$table = new \CodeIgniter\View\Table();
$table->setFooting($footing);
echo $table->generate($data);

echo '<p>Update template footer to use {th} rather then {td}.</p>';
$template = [
	'footing_cell_start' => '<th>',
	'footing_cell_end' => '</th>'
];
$table = new \CodeIgniter\View\Table($template);
$table->setFooting($footing);
echo $table->generate($data);
```

### Expected Output

I expect cells in the footer to be accept attributes in the same manner as cells in the table body or table header. 

### Anything else?

line 365 of System\View\Table
`$temp = str_replace('<th', '<th ' . $key . '="' . $val . '"', $temp);`

{th} should be {td}, if the template is default

---

But, of course, what if the template has been altered by the App? This replacement method is only going to work if the template 'table_footer_start' starts with '<th'.

For now... make it work for the _default_ template. If people change the template, they probably will not use the extra attributes.

### PR DESCRIPTION
**Pull request by ddevsr**
**Description**
Fixed #6507
Supersedes #6511

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by ddevsr** [comment]
@kenjis Sorry, i am newbie to using command rebase mechanism. Now can use GPG Signing at Home PC

File: system/View/Table.php [hunk 1]
```diff
@@ -300,9 +300,14 @@ public function generate($tableData = null)
             foreach ($this->heading as $heading) {
                 $temp = $this->template['heading_cell_start'];
 
+                // Check tag with td or th
+                // If not will return null
+                preg_match('/(<)(td|th)/i', $temp, $valueHead);
+                $strHeading = count($valueHead) === 3 ? $valueHead[1] . $valueHead[2] : null;
```
Comment (kenjis) [inline system/View/Table.php:306]: I think we only need to check it once.

```diff
--- a/system/View/Table.php
+++ b/system/View/Table.php
@@ -293,6 +293,10 @@ class Table
             $out .= '<caption>' . $this->caption . '</caption>' . $this->newline;
         }
 
+        // Check if there is a th or td tag, null if not
+        preg_match('/(<)(td|th)/i', $this->template['heading_cell_start'], $valueHead);
+        $strHeading = count($valueHead) === 3 ? $valueHead[1] . $valueHead[2] : null;
+
         // Is there a table heading to display?
         if (! empty($this->heading)) {
             $out .= $this->template['thead_open'] . $this->newline . $this->template['heading_row_start'] . $this->newline;
@@ -300,11 +304,6 @@ class Table
             foreach ($this->heading as $heading) {
                 $temp = $this->template['heading_cell_start'];
 
-                // Check tag with td or th
-                // If not will return null
-                preg_match('/(<)(td|th)/i', $temp, $valueHead);
-                $strHeading = count($valueHead) === 3 ? $valueHead[1] . $valueHead[2] : null;
-
                 foreach ($heading as $key => $val) {
                     if ($key !== 'data') {
                         $temp = $strHeading ? str_replace($strHeading, $strHeading . ' ' . $key . '="' . $val . '"', $temp) : $temp;
```

File: system/View/Table.php [hunk 2]
```diff
@@ -360,9 +365,14 @@ public function generate($tableData = null)
             foreach ($this->footing as $footing) {
                 $temp = $this->template['footing_cell_start'];
 
+                // Check tag with td or th
+                // If not will return null
+                preg_match('/(<)(td|th)/i', $temp, $valueFoot);
+                $strFooting = count($valueFoot) === 3 ? $valueFoot[1] . $valueFoot[2] : null;
```
Comment (kenjis) [inline system/View/Table.php:371]: Same as heading.

File: tests/system/View/TableTest.php [hunk 3]
```diff
@@ -22,10 +22,16 @@
...
+    private string $styleTableOne = 'background:#F99;text-align:right;width:5em;';
+    private string $styleTableTwo = 'background:cyan;color:white;text-align:right;width:5em;';
 
     protected function setUp(): void
     {
         $this->table = new MockTable();
+
+        // Reset Header & Footer
+        $this->table->heading = [];
+        $this->table->footing = [];
```
Comment (kenjis) [inline tests/system/View/TableTest.php:34]: Remove these lines.
`$this->table` is a new instance, so they are not needed.

**Review by kenjis — APPROVED** [review]
LGTM!

File: system/View/Table.php [hunk 4]
```diff
@@ -295,14 +295,17 @@ public function generate($tableData = null)
 
         // Is there a table heading to display?
         if (! empty($this->heading)) {
+            preg_match('/(<)(td|th)/i', $this->template['heading_cell_start'], $valueHead);
+            $strHeading = count($valueHead) === 3 ? $valueHead[1] . $valueHead[2] : null;
```
Comment (paulbalandan) [inline system/View/Table.php:299]: Why do you need the `count()` evaluation? If the string matches the pattern, then `$valueHead[0]` should be the same as `$valueHead[1] . $valueHead[2]`. Isn't it?
https://3v4l.org/6rSB2

File: system/View/Table.php [hunk 5]
```diff
@@ -295,14 +295,17 @@ public function generate($tableData = null)
...
+
             $out .= $this->template['thead_open'] . $this->newline . $this->template['heading_row_start'] . $this->newline;
 
             foreach ($this->heading as $heading) {
                 $temp = $this->template['heading_cell_start'];
 
                 foreach ($heading as $key => $val) {
                     if ($key !== 'data') {
-                        $temp = str_replace('<th', '<th ' . $key . '="' . $val . '"', $temp);
+                        $temp = $strHeading ? str_replace($strHeading, $strHeading . ' ' . $key . '="' . $val . '"', $temp) : $temp;
```
Comment (paulbalandan) [inline system/View/Table.php:308]: This ternary condition can be moved up to the `if` above to check if `$strHeading !== null`

File: system/View/Table.php [hunk 6]
```diff
@@ -355,14 +358,18 @@ public function generate($tableData = null)
 
         // Any table footing to display?
         if (! empty($this->footing)) {
+            // Check if there is a th or td tag, null if not
+            preg_match('/(<)(td|th)/i', $this->template['footing_cell_start'], $valueFoot);
+            $strFooting = count($valueFoot) === 3 ? $valueFoot[1] . $valueFoot[2] : null;
```
Comment (paulbalandan) [inline system/View/Table.php:363]: Same with 1st comment

File: system/View/Table.php [hunk 7]
```diff
@@ -355,14 +358,18 @@ public function generate($tableData = null)
...
+
             $out .= $this->template['tfoot_open'] . $this->newline . $this->template['footing_row_start'] . $this->newline;
 
             foreach ($this->footing as $footing) {
                 $temp = $this->template['footing_cell_start'];
 
                 foreach ($footing as $key => $val) {
                     if ($key !== 'data') {
-                        $temp = str_replace('<th', '<th ' . $key . '="' . $val . '"', $temp);
+                        $temp = $strFooting ? str_replace($strFooting, $strFooting . ' ' . $key . '="' . $val . '"', $temp) : $temp;
```
Comment (paulbalandan) [inline system/View/Table.php:372]: Same with 2nd comment

File: system/View/Table.php (hunk 4 above)
Comment (ddevsr) [reply system/View/Table.php:299]: With pattern now, i just checked `<thd>` or `<tdh>` can be accept (see [https://3v4l.org/KvjvH](https://3v4l.org/KvjvH)), i can update with `'/(<)(td|th)(>)/i'` pattern (see [https://3v4l.org/ZB3Nn](https://3v4l.org/ZB3Nn)). With new pattern i need `$valueHead[1] . $valueHead[2]`

Comment (ddevsr) [reply system/View/Table.php:299]:
```php
--- a/system/View/Table.php
+++ b/system/View/Table.php
@@ -295,8 +295,8 @@ class Table

         // Is there a table heading to display?
         if (! empty($this->heading)) {
-            preg_match('/(<)(td|th)/i', $this->template['heading_cell_start'], $valueHead);
-            $strHeading = count($valueHead) === 3 ? $valueHead[1] . $valueHead[2] : null;
+            preg_match('/(<)(td|th)(>)/i', $this->template['heading_cell_start'], $valueHead);
+            $strHeading = count($valueHead) === 4 ? $valueHead[1] . $valueHead[2] : null;

             $out .= $this->template['thead_open'] . $this->newline . $this->template['heading_row_start'] . $this->newline;

@@ -304,8 +304,8 @@ class Table
                 $temp = $this->template['heading_cell_start'];

                 foreach ($heading as $key => $val) {
-                    if ($key !== 'data') {
-                        $temp = $strHeading ? str_replace($strHeading, $strHeading . ' ' . $key . '="' . $val . '"', $temp) : $temp;
+                    if ($key !== 'data' && $strHeading !== null) {
+                        $temp = str_replace($strHeading, $strHeading . ' ' . $key . '="' . $val . '"', $temp);
                     }
                 }

@@ -359,8 +359,8 @@ class Table
         // Any table footing to display?
         if (! empty($this->footing)) {
             // Check if there is a th or td tag, null if not
-            preg_match('/(<)(td|th)/i', $this->template['footing_cell_start'], $valueFoot);
-            $strFooting = count($valueFoot) === 3 ? $valueFoot[1] . $valueFoot[2] : null;
+            preg_match('/(<)(td|th)(>)/i', $this->template['footing_cell_start'], $valueFoot);
+            $strFooting = count($valueFoot) === 4 ? $valueFoot[1] . $valueFoot[2] : null;

             $out .= $this->template['tfoot_open'] . $this->newline . $this->template['footing_row_start'] . $this->newline;

@@ -368,8 +368,8 @@ class Table
                 $temp = $this->template['footing_cell_start'];

                 foreach ($footing as $key => $val) {
-                    if ($key !== 'data') {
-                        $temp = $strFooting ? str_replace($strFooting, $strFooting . ' ' . $key . '="' . $val . '"', $temp) : $temp;
+                    if ($key !== 'data' && $strFooting !== null) {
+                        $temp = str_replace($strFooting, $strFooting . ' ' . $key . '="' . $val . '"', $temp);
                     }
                 }
```

Comment (paulbalandan) [reply system/View/Table.php:299]: I think I didn't get my point through.

My point is in this pattern `/(<)(td|th)/i`, why use `$matches[1] . $matches[2]` when you can achieve the same with `$matches[0]`?

In https://3v4l.org/KvjvH, the `0`-key is `<td`, which is the concat of `1`-key and `2`-key.
```
array (
  0 => '<td',
  1 => '<',
  2 => 'td',
)
```

Now, on your point that the pattern will also match `<tdh>`, you can use lookahead to whitespace or `>`. Like `/(<)(td|th)(?=\h|>)/i`. See https://3v4l.org/2fm0O

Comment (ddevsr) [reply system/View/Table.php:299]: Thank you suggest `/(<)(td|th)(?=\h|>)/i` pattern, when I access $valueHead[0] directly return error

```
There were 2 errors:

1) CodeIgniter\View\TableTest::testGenerateInvalidHeadingFooting
Undefined offset: 0

D:\Project\laragon\www\ci4\system\View\Table.php:299
D:\Project\laragon\www\ci4\tests\system\View\TableTest.php:568
phpvfscomposer://D:\Project\laragon\www\ci4\vendor\phpunit\phpunit\phpunit:97

2) CodeIgniter\View\TableTest::testGenerateInvalidHeadingFootingHTML
Undefined offset: 0

D:\Project\laragon\www\ci4\system\View\Table.php:299
D:\Project\laragon\www\ci4\tests\system\View\TableTest.php:617
phpvfscomposer://D:\Project\laragon\www\ci4\vendor\phpunit\phpunit\phpunit:97

ERRORS!
Tests: 26, Assertions: 78, Errors: 2.
```

That's why I use `$valueHead[1] . $valueHead[2]`
image

Comment (ddevsr) [reply system/View/Table.php]: No error

**Review by paulbalandan — APPROVED** [review]

**Review by kenjis — APPROVED** [review]

**Pull request merged**